File: extension/substrate/bindingDelivery.ts
// Warm-door (TS-extension) delivery of resolved skill bindings — the in-session twin of the cold
// door (perk/substrate/binding_delivery.py). Both planes render the SAME resolved overlay (defaults
// ⊕ user bindings; `nudge` -> a pointer line, `transclude` -> the inlined skill body) under the
// SAME header literal. The cold door appends it to a launch's initial prompt; this module renders
// it at two WARM surfaces:
//
//   Mechanism A — stage triggers: a `before_agent_start` handler (mirroring planMode.ts /
//   objectiveAuthor.ts) injects the launched stage's bindings as a hidden context message — the
//   delivery path for `stage:plan`'s `perk-plan` pointer, since a cold `perk plan` launches idle
//   (no initial prompt to augment).
//   Mechanism B — `bindingSuffix()` is appended into the guidance of perk's warm slash-commands so
//   each self-delivers its pointer (a warm `/objective-plan` run outside a stage:objective-plan
//   session gets none from Mechanism A).
//
// This is the SINGLE delivery path for perk's own nudges. Delivery NEVER double-delivers: the
// cold↔warm dedup marker is `BINDING_HEADER` itself — the cold door's initial prompt and every warm
// injection carry it, so Mechanism A injects ONLY when nothing on the branch already carries the
// header (idempotent across turns/reloads; after compaction drops the original it re-delivers).
//
// LBYL throughout: a missing/unreadable transclude target degrades to the nudge pointer with a
// loud-but-non-fatal warning, never throws, never blocks a turn. Resolver shape `issues` are NOT
// surfaced warm (the cold launch + doctor own them); only the transclude `warnings` are.

import { existsSync, readFileSync } from "node:fs";
import { join } from "node:path";
import type { ExtensionAPI } from "@earendil-works/pi-coding-agent";
import { loadDefaultBindings, resolveBindings, type SkillBinding } from "./bindings.ts";
import { loadPerkConfig } from "./config.ts";
import { type BranchEntry, branchOf, rebuildWorkflowState } from "./workflowState.ts";

/**
 * The cross-plane dedup marker AND render header. MUST stay byte-identical to the Python cold
 * door's `_HEADER` (perk/substrate/binding_delivery.py) — both planes render under it so a cold launch and a
 * warm injection never double-deliver. Pinned by a literal test in both planes (§8.9).
 */
export const BINDING_HEADER = "The following skill binding(s) apply here:";

/** The hidden context customType carrying a warm-injected stage-binding render (Mechanism A). */
export const BINDING_CONTEXT_TYPE = "perk:binding-context";

const SKILLS_SUBDIR = join(".agents", "skills");
const SKILL_FILENAME = "SKILL.md";

/** The rendered warm delivery for one trigger: the prompt fragment (or `null`) + any warnings. */
export interface BindingRender {
  text: string | null;
  warnings: string[];
}

/**
 * The full resolved bindings: the shipped defaults ⊕ the user overlay. The TS twin of cold's
 * `resolve_bindings(...).bindings` (perk/substrate/binding_delivery.py) — delivers the defaults too
 * (perk's own nudges are no longer hardcoded), so there is no longer a default subtraction.
 */
export function resolvedBindings(cwd: string): SkillBinding[] {
  return resolveBindings(loadPerkConfig(cwd).bindings, loadDefaultBindings()).bindings;
}

/**
 * Render the resolved bindings matching `trigger` into a header-joined fragment (or `null` when
 * none match). `nudge` renders a `Follow the \`<skill>\` skill.` pointer; `transclude` inlines
 * `.agents/skills/<skill>/SKILL.md` (frontmatter stripped), degrading to the nudge pointer with a
 * loud-but-non-fatal warning when the file is absent/unreadable. Pure but for the LBYL file read.
 */
export function renderBindings(cwd: string, trigger: string): BindingRender {
  const mine = resolvedBindings(cwd).filter((binding) => binding.trigger === trigger);
  const warnings: string[] = [];
  const parts: string[] = [];
  for (const binding of mine) {
    if (binding.mode === "transclude") {
      const body = readSkillBody(cwd, binding.skill);
      if (body !== null) {
        parts.push(`Skill \`${binding.skill}\` (inlined for \`${binding.trigger}\`):\n\n${body}`);
        continue;
      }
      warnings.push(
        `skill binding: transclude target for \`${binding.skill}\` not found under ` +
          `${SKILLS_SUBDIR}/${binding.skill}/${SKILL_FILENAME} — falling back to a pointer.`,
      );
    } else if (!skillInstalled(cwd, binding.skill)) {
      // The nudge mirror of the transclude warning: a binding to a skill that is
      // not installed is reported loud-but-non-fatal, never silently delivered. The pointer is
      // still emitted so the model gets the nudge.
      warnings.push(
        `skill binding: skill \`${binding.skill}\` for \`${binding.trigger}\` is not installed ` +
          `under ${SKILLS_SUBDIR}/${binding.skill}/${SKILL_FILENAME} — the pointer may dangle.`,
      );
    }
    parts.push(`Follow the \`${binding.skill}\` skill.`);
  }
  const text = parts.length > 0 ? [BINDING_HEADER, ...parts].join("\n\n") : null;
  return { text, warnings };
}

/**
 * The Mechanism-B suffix: the rendered bindings for `trigger` to append into a warm command's
 * guidance (empty string when none match). Every perk warm slash-command self-delivers its pointer
 * this way, by `stage:<id>` or `command:<id>`. A leading blank line keeps it
 * visually distinct from the guidance it follows. Any render warnings (missing transclude target or
 * uninstalled nudge skill) are `console.error`-ed loud-but-non-fatal — the nudge
 * fallback still reaches the model, but the misconfiguration is no longer surfaced silently.
 */
export function bindingSuffix(cwd: string, trigger: string): string {
  const { text, warnings } = renderBindings(cwd, trigger);
  for (const warning of warnings) console.error(`perk: ${warning}`);
  return text ? `\n\n${text}` : "";
}

/** Whether `.agents/skills/<skill>/SKILL.md` exists under `cwd` (the warm delivery read path). */
function skillInstalled(cwd: string, skill: string): boolean {
  return existsSync(join(cwd, SKILLS_SUBDIR, skill, SKILL_FILENAME));
}

/** Read `.agents/skills/<skill>/SKILL.md` (frontmatter stripped); `null` if absent/unreadable. */
function readSkillBody(cwd: string, skill: string): string | null {
  const path = join(cwd, SKILLS_SUBDIR, skill, SKILL_FILENAME);
  if (!existsSync(path)) return null;
  try {
    // Normalize CRLF/CR before stripping (the miniJinja.ts pattern): Node's readFileSync keeps
    // `\r\n` where Python's read_text() normalizes, so without this a CRLF checkout would defeat
    // the `---\n` frontmatter check AND break the cross-plane byte parity pinned by
    // tests/test_binding_render_parity.py.
    return stripFrontmatter(readFileSync(path, "utf8").replace(/\r\n?/g, "\n"));
  } catch {
    return null;
  }
}

/** Drop a leading `---`-delimited YAML frontmatter block; return the body stripped. */
function stripFrontmatter(text: string): string {
  if (!text.startsWith("---\n")) return text;
  const lines = text.split("\n");
  for (let i = 1; i < lines.length; i++) {
    if (lines[i] === "---")
      return lines
        .slice(i + 1)
        .join("\n")
        .trim();
  }
  return text; // no closing delimiter — leave the text unchanged
}

/**
 * Whether anything on the branch already carries `BINDING_HEADER` — the cold launch's initial
 * prompt OR a prior warm injection. Serializing each entry is the robust, shape-agnostic scan: the
 * header is a distinctive literal, so a substring hit means "already delivered on this branch".
 */
function branchHasHeader(branch: readonly BranchEntry[]): boolean {
  return branch.some((entry) => JSON.stringify(entry).includes(BINDING_HEADER));
}

/** The launched stage's `stage:<id>` render, or `null` when there is no stage / nothing matches. */
function activeStageRender(cwd: string, branch: readonly BranchEntry[]): BindingRender | null {
  const stage = rebuildWorkflowState(branch).stage;
  if (!stage) return null;
  return renderBindings(cwd, `stage:${stage}`);
}

/**
 * Register warm-door binding delivery: Mechanism A's dedup-guarded `before_agent_start` injection
 * plus a `context` strip mirroring planMode.ts / objectiveAuthor.ts (keep while the stage's
 * bindings are live; strip the stale custom otherwise). Inert when nothing matches the stage;
 * never throws. Mechanism B (`bindingSuffix`) is wired by the command modules themselves.
 */
export function registerBindingDelivery(pi: ExtensionAPI): void {
  // Mechanism A — inject the launched stage's resolved bindings as a hidden context message,
  // but ONLY when no entry on the branch already carries BINDING_HEADER (the cold door's initial
  // prompt or a prior warm inject) — the cold↔warm idempotency guard.
  pi.on("before_agent_start", async (_event, ctx) => {
    const branch = branchOf(ctx);
    const rendered = activeStageRender(ctx.cwd, branch);
    if (rendered === null || rendered.text === null) return;
    if (branchHasHeader(branch)) return;
    for (const warning of rendered.warnings) console.error(`perk: ${warning}`);
    return {
      message: {
        customType: BINDING_CONTEXT_TYPE,
        content: rendered.text,
        display: false,
      },
    };
  });

  // Strip a STALE binding-context custom from the model window when the current stage no longer
  // renders bindings (stage changed, or the overlay was removed) — the same hygiene planMode /
  // objectiveAuthor apply to their authoring contexts. While the stage IS live we keep it (the
  // model must see the nudge), and the dedup above relies on it persisting on the branch.
  //
  // Deliberately NARROWER than planMode: it strips ONLY the BINDING_CONTEXT_TYPE custom, never a
  // user message carrying the header — a cold launch's initial prompt legitimately carries
  // BINDING_HEADER and must survive in context.
  pi.on("context", async (event, ctx) => {
    const branch = branchOf(ctx);
    const rendered = activeStageRender(ctx.cwd, branch);
    if (rendered !== null && rendered.text !== null) return;
    return {
      messages: event.messages.filter((m) => {
        const msg = m as { customType?: string };
        return msg.customType !== BINDING_CONTEXT_TYPE;
      }),
    };
  });
}

File: src/perk/substrate/binding_delivery.py
"""Cold-door (Python-plane) delivery of the resolved skill-binding overlay (§8.9).

When `perk` launches a `pi` session, this renders the **full resolved** bindings whose trigger
matches the launch into the session's initial prompt — `nudge` as a pointer line carrying the
skill's read path (perk's stage skills are hidden from the ambient system prompt, so the pointer
must name where the body lives), `transclude` as the inlined skill body. The mechanism is perk's
**single delivery path** for its own nudges: the shipped defaults carry no hardcoded per-skill
nudge string and are delivered here.
The warm door (the TS extension) is the in-session twin; target-existence validation (is
`stage:x` a real stage? is the skill installed?) stays `doctor`.

LBYL throughout (dignified-python): a missing/unreadable transclude target is surfaced as a
loud-but-non-fatal warning and degrades to the nudge pointer — never raises, never blocks a
launch.
"""

from dataclasses import dataclass
from pathlib import Path

from perk.substrate.bindings import (
    SKILL_FILENAME,
    SKILLS_DIR,
    Binding,
    is_skill_installed,
    load_bindings,
    resolve_bindings,
)
from perk.substrate.registry import Issue

_HEADER = "The following skill binding(s) apply here:"


@dataclass(frozen=True)
class ColdBindingDelivery:
    """The rendered cold-door delivery for one launch trigger.

    ``text`` is the prompt fragment to append (``None`` when nothing matches the trigger);
    ``issues`` carries the resolver's shape/duplicate findings; ``warnings`` carries delivery
    warnings (e.g. a missing transclude target) — both surfaced loud-but-non-fatal by the caller.
    """

    text: str | None
    issues: list[Issue]
    warnings: list[str]


def render_cold_bindings(
    user_bindings: list[Binding],
    repo_root: Path,
    trigger: str,
    *,
    defaults: list[Binding] | None = None,
) -> ColdBindingDelivery:
    """Render the full resolved bindings matching ``trigger`` into a prompt fragment.

    The resolved set is the shipped defaults ⊕ the user overlay (perk's own nudges are not
    hardcoded — the defaults are delivered here too). Only bindings whose trigger equals
    ``trigger`` are delivered. Resolver issues and delivery warnings are collected (never raised)
    for the caller to surface loud-but-non-fatal.
    """
    if defaults is None:
        defaults = load_bindings().bindings
    resolved = resolve_bindings(user_bindings, defaults=defaults)
    mine = [b for b in resolved.bindings if b.trigger == trigger]

    warnings: list[str] = []
    parts: list[str] = []
    for binding in mine:
        if binding.mode == "transclude":
            body = _read_skill_body(repo_root, binding.skill)
            if body is not None:
                parts.append(
                    f"Skill `{binding.skill}` (inlined for `{binding.trigger}`):\n\n{body}"
                )
                continue
            warnings.append(
                f"skill binding: transclude target for `{binding.skill}` not found under "
                f"{SKILLS_DIR}/{binding.skill}/{SKILL_FILENAME} — falling back to a pointer."
            )
        elif not is_skill_installed(repo_root, binding.skill):
            # The nudge mirror of the transclude warning (D6): a binding to a skill that
            # is not installed is reported loud-but-non-fatal, never silently delivered. The pointer
            # is still emitted so the model gets the nudge.
            warnings.append(
                f"skill binding: skill `{binding.skill}` for `{binding.trigger}` is not installed "
                f"under {SKILLS_DIR}/{binding.skill}/{SKILL_FILENAME} — the pointer may dangle."
            )
        parts.append(
            f"Follow the `{binding.skill}` skill "
            f"(read `{SKILLS_DIR}/{binding.skill}/{SKILL_FILENAME}`)."
        )

    text = "\n\n".join([_HEADER, *parts]) if parts else None
    return ColdBindingDelivery(text=text, issues=list(resolved.issues), warnings=warnings)


def _read_skill_body(repo_root: Path, skill: str) -> str | None:
    """Read ``.agents/skills/<skill>/SKILL.md`` (frontmatter stripped); ``None`` if absent or
    unreadable — an undecodable/unreadable file degrades to the same pointer fallback as
    absence (parity with ``bindingDelivery.ts::readSkillBody``), never a crashed launch."""
    path = repo_root / SKILLS_DIR / skill / SKILL_FILENAME
    if not path.is_file():
        return None
    try:
        return _strip_frontmatter(path.read_text(encoding="utf-8"))
    except (OSError, UnicodeDecodeError):
        return None


def _strip_frontmatter(text: str) -> str:
    """Drop a leading ``---``-delimited YAML frontmatter block; return the body stripped.

    If the text does not open with a frontmatter block, it is returned unchanged.
    """
    if not text.startswith("---\n"):
        return text
    lines = text.split("\n")
    for i in range(1, len(lines)):
        if lines[i] == "---":
            return "\n".join(lines[i + 1 :]).strip()
    return text  # no closing delimiter — leave the text unchanged
